CodingStyle

```diff
@@ -129,3 +129,28 @@ type conversions.
 * Other C++ Features > Avoid Default Arguments:
 
     They obscure the interface.
+
+
+Python code
+-----------
+
+For all python code, PEP-8 should be observed:
+
+    https://www.python.org/dev/peps/pep-0008/
+
+Existing code should not be converted, much like what we do for both C++ and
+C, unless it is being changed significantly.
```

Comment:
This half-baked enforcement means that it will be difficult to detect if a change is compliant unless it is super obvious. 

PEP8 is not super obvious all the time which is why most rely on a text-editor with a linter that shows these in real time.

This means we can't have an automated check because it will never pass from the get go. How could one possibly identify what is wrong with a proposed change if the errors are mixed with other parts? 

In other Python projects we made the effort to clean up and then enforce.

Reply:
@alfredodeza any suggestions how we can make this work better then? Ideally, I would love to have some form of coding style guidelines for code to come (e.g., mgr modules), but would love even more to do this the right way :)

Reply:
@alfredodeza I'm guessing enforcing and forcing things to be cleaned up would then address your point? Any suggestion on a linter to enforce PEP8 (or something else)? (I'm not sure whether the depth - or lack thereof - of my knowledge on this topic is fully transpiring ;) )

Reply:
I mean...this isn't any enforcement at all, and isn't trying to be; it's just a style README

AFAIK we don't enforce C++ or shell standards by and large

Reply:
What is the use of a "style README" that can't be verified, and therefore it can't be asked for correction? If a PR comes along with non-pep8 standards, then is it OK then to just merge because we don't really enforce it?

This is too ambiguous, we either want code in pep8 style or not. If we don't want to enforce it then I don't see the need of adding this to a README

Reply:
Even if we are not enforcing it **now**, it doesn't mean we can't establish a guideline. I'd rather establish a guideline we can point people to, rather than having arguments over whether lines should be this length or the other (because there's no defined policy), and try to eventually achieve a code base that we can strictly enforce via linting.

Reply:
What's the use of an opinion that's not a law?

Reply:
How painful would it be to make the existing code conform?

Reply:
While I would like a jenkins job to verify PEP8 as much as the next person, specifying it in our style guide is still useful without the verification -- it enables reviewers to ask contributors to get their code reasonably close to PEP8 before merging.  Especially if we're planning to add enforcement in the future, it makes life easier if we communicate this now.

As for converting existing code -- most of the mgr code is fairly close to pep8 (i.e. it doesn't light up like crazy in pycharm), and I wouldn't mind accepting patches that take it the rest of the way there, once we've decided how strict we want to be.  We should concentrate on the mgr code (where all the new stuff is happening) rather than the rest of pybind, where it probably matters less.

Reply:
:+1:

Reply:
So, it is reasonable to leave it as is, or would you like me to reword some of it?

Reply:
> What's the use of an opinion that's not a law?

IMHO, the benefit of having a guideline outweighs the "half-baked enforcement". we do have existing code not conforming to the C++ guideline we are enforcing (please allow me use this word =), but this does not prevent us from having a code style guideline, just because we don't have an automatic machinery to detect the inconsistencies, for instance, the trailing spaces in C++ source code.

i don't think we can fix all of them problem in this single PR.

- for enforcing the C++ coding style, we could have a .clang-format, and suggest developers to install a pre-commit hook.
- for enforcing the PEP8: we could use pep8/flake8 in the pre-commit hook also.

on the jenkins side, there is more work to do.  assuming the levels of conforming to code style guideline can be compared, we could keep track of the current status of "goodness" in master in a repo (could be a github repo), and only accept the PR which does not decrease the level of conformity.

yeah, all of these need more love and efforts. but we can do it by piecemeal.